File: src/client/postback-engine.ts
import type { HttpClient } from './http-client';
import type { AspNetWebFormData } from '../types/session';
import { extractAspNetFormData, buildPostbackData } from '../parsers/aspnet-parser';
import { VULMS_BASE_URL } from '../constants/urls';
import { noopLogger, type Logger } from '../utils/logger';
import { isImageButton, buildImageButtonFields, extractPageTitle } from '../utils/navigation';
import { validateLecturePage, validateQuizPage, validateAssignmentPage, validateGDBPage, type PageValidationResult } from '../utils/validation';

export interface PostBackOptions {
  page: string;
  eventTarget?: string;
  eventArgument?: string;
  extraFields?: Record<string, string>;
  refreshFormState?: boolean;
}

export class PostBackEngine {
  private httpClient!: HttpClient;
  private lastFormData: AspNetWebFormData | null = null;
  private lastPage: string | null = null;
  private debug!: Logger;

  constructor(httpClient: HttpClient, debug: Logger = noopLogger) {
    this.httpClient = httpClient;
    this.debug = debug;
  }

  async performPostBack(options: PostBackOptions): Promise<string> {
    const target = options.eventTarget || 'none';
    const isImgBtn = isImageButton(target);

    this.debug.debug(`PostBack: ${options.page} [${target}] type=${isImgBtn ? 'image-button' : 'standard'}`);

    const needsRefresh =
      options.refreshFormState || !this.lastFormData || this.lastPage !== options.page;

    if (needsRefresh) {
      this.debug.debug(`  Fetching form state from ${options.page}`);
      const pageHtml = await this.httpClient.get({ path: options.page });
      this.lastFormData = extractAspNetFormData(pageHtml);
      this.lastPage = options.page;
      this.debug.debug(`  VIEWSTATE length: ${this.lastFormData.__VIEWSTATE.length}`);
    }

    const fields: Record<string, string> = {};

    if (isImgBtn) {
      const imgFields = buildImageButtonFields(target);
      for (const [k, v] of Object.entries(imgFields)) {
        fields[k] = v;
      }
      this.debug.debug(`  Image button: ${target} + ${target}.x/y`);
    } else {
      if (options.eventTarget) {
        fields.__EVENTTARGET = options.eventTarget;
        this.debug.debug(`  __EVENTTARGET=${options.eventTarget}`);
      }
    }

    if (options.eventArgument) fields.__EVENTARGUMENT = options.eventArgument;

    const data = buildPostbackData(this.lastFormData!, {
      ...fields,
      ...options.extraFields,
    });

    const html = await this.httpClient.post({
      path: options.page,
      data,
      referer: `${VULMS_BASE_URL}${options.page}`,
    });

    const pageTitle = extractPageTitle(html);
    this.debug.debug(`  Response: title="${pageTitle}" length=${html.length}`);

    try {
      this.lastFormData = extractAspNetFormData(html);
      this.lastPage = options.page;
      this.debug.debug(`  Response OK, VIEWSTATE length: ${this.lastFormData.__VIEWSTATE.length}`);
    } catch (e) {
      this.debug.warn(`  Failed to extract form state from response: ${(e as Error).message}`);
      this.lastFormData = null;
      this.lastPage = null;
    }

    return html;
  }

  async performNavigation(
    page: string,
    eventTarget: string,
    expectedRepeater: 'quiz' | 'assignment' | 'gdb' | 'lecture',
    courseCode?: string,
    refreshState?: boolean,
  ): Promise<string> {
    const isImg = isImageButton(eventTarget);
    const tag = '[POSTBACK]';
    const target = isImg ? eventTarget : (eventTarget || 'none');

    this.debug.debug(`${tag} target=${target} page=${page} isImageButton=${isImg} expect=${expectedRepeater} refreshState=${refreshState}`);

    const startTime = Date.now();
    const html = await this.performPostBack({ page, eventTarget, refreshFormState: refreshState });
    const elapsed = Date.now() - startTime;

    const validation = this.validatePage(html, expectedRepeater);
    const title = extractPageTitle(html);
    const formAction = this.extractFormAction(html);

    this.debug.debug(`[NAVIGATION RESULT] course=${courseCode || 'unknown'} state=${validation.state} url=${formAction} htmlLength=${html.length} title="${title}" elapsed=${elapsed}ms indicators=[${validation.indicators.join(', ')}]`);

    if (validation.state === 'INVALID') {
      const errorMsg = `Navigation validation failed for ${courseCode || 'unknown'}: expected=${expectedRepeater}, state=${validation.state}, title="${title}", missing=${validation.missingExpected || 'unknown'}`;
      this.debug.warn(`[NAVIGATION FAILED] ${errorMsg}`);
      throw new Error(errorMsg);
    }

    return html;
  }

  private validatePage(html: string, type: 'quiz' | 'assignment' | 'gdb' | 'lecture'): PageValidationResult {
    switch (type) {
      case 'quiz': return validateQuizPage(html);
      case 'assignment': return validateAssignmentPage(html);
      case 'gdb': return validateGDBPage(html);
      case 'lecture': return validateLecturePage(html);
    }
  }

  private extractFormAction(html: string): string {
    const match = html.match(/<form[^>]*action="([^"]*)"/i);
    return match ? match[1] : '';
  }

  async fetchWithFormState(page: string): Promise<string> {
    this.debug.debug(`fetchWithFormState: ${page}`);
    const html = await this.httpClient.get({ path: page });
    this.lastFormData = extractAspNetFormData(html);
    this.lastPage = page;
    return html;
  }

  clearState(): void {
    this.lastFormData = null;
    this.lastPage = null;
    this.debug.debug('PostBack state cleared');
  }

  hasState(): boolean {
    return this.lastFormData !== null && this.lastPage !== null;
  }
}

File: vulms_dart/lib/src/client/postback_engine.dart
import '../models/session.dart';
import '../networking/vulms_http_client.dart';
import '../parsers/aspnet_parser.dart';
import '../utils/logger.dart';
import '../utils/constants.dart';

/// Handles ASP.NET PostBack navigation.
///
/// Manages form state (__VIEWSTATE, __EVENTVALIDATION) across requests
/// and supports both standard PostBack and image button PostBack.
class PostBackEngine {
  final VulmsHttpClient _httpClient;
  final VulmsLogger _logger;
  AspNetFormData? _lastFormData;
  String? _lastPage;

  PostBackEngine({
    required VulmsHttpClient httpClient,
    required VulmsLogger logger,
  })  : _httpClient = httpClient,
        _logger = logger;

  /// Perform a PostBack request.
  ///
  /// [page] is the ASP.NET page path.
  /// [eventTarget] is the __EVENTTARGET value (control that triggered the postback).
  /// [extraFields] are additional form fields to include.
  /// [refreshFormState] forces re-fetching the form state before posting.
  Future<String> performPostBack({
    required String page,
    String? eventTarget,
    Map<String, String>? extraFields,
    bool refreshFormState = false,
  }) async {
    final isImgBtn = _isImageButton(eventTarget ?? '');
    _logger.debug(
        'PostBack: $page [${eventTarget ?? "none"}] type=${isImgBtn ? "image-button" : "standard"}');

    final needsRefresh = refreshFormState ||
        _lastFormData == null ||
        _lastPage != page;

    if (needsRefresh) {
      _logger.debug('  Fetching form state from $page');
      final pageHtml = await _httpClient.get(path: page);
      _lastFormData = AspNetParser.extractFormData(pageHtml);
      _lastPage = page;
      _logger.debug('  VIEWSTATE length: ${_lastFormData!.viewState.length}');
    }

    final fields = <String, String>{};

    if (isImgBtn) {
      fields[eventTarget!] = 'Submit';
      fields['${eventTarget}.x'] = '10';
      fields['${eventTarget}.y'] = '10';
      _logger.debug('  Image button: $eventTarget + x/y');
    } else {
      if (eventTarget != null) {
        fields['__EVENTTARGET'] = eventTarget;
        _logger.debug('  __EVENTTARGET=$eventTarget');
      }
    }

    final data = AspNetParser.buildPostbackData(
      _lastFormData!,
      extraFields: {...fields, ...?extraFields},
    );

    final html = await _httpClient.post(
      path: page,
      data: data,
      referer: '${VulmsUrls.baseUrl}$page',
    );

    _logger.debug('  Response length: ${html.length}');

    try {
      _lastFormData = AspNetParser.extractFormData(html);
      _lastPage = page;
      _logger.debug('  Response OK, VIEWSTATE length: ${_lastFormData!.viewState.length}');
    } catch (e) {
      _logger.warn('  Failed to extract form state from response: $e');
      _lastFormData = null;
      _lastPage = null;
    }

    return html;
  }

  /// Fetch a page and store its form state.
  Future<String> fetchWithFormState(String page) async {
    _logger.debug('fetchWithFormState: $page');
    final html = await _httpClient.get(path: page);
    _lastFormData = AspNetParser.extractFormData(html);
    _lastPage = page;
    return html;
  }

  /// Clear stored form state.
  void clearState() {
    _lastFormData = null;
    _lastPage = null;
    _logger.debug('PostBack state cleared');
  }

  /// Whether form state is currently stored.
  bool hasState() => _lastFormData != null && _lastPage != null;

  bool _isImageButton(String name) {
    return name.contains('ibtn') || name.contains(r'$ibtn');
  }
}
